### ISSUE TITLE
Set buff size based on aspect ratio, dpi, and window size

### ISSUE THREAD

**Issue by yt-fido** [yt-project/yt#1021]
Originally reported by: **Matt Turk (Bitbucket: [MatthewTurk](https://bitbucket.org/MatthewTurk), GitHub: [MatthewTurk](https://github.com/MatthewTurk))**

----------------------------------------

Right now, it seems like our `buff_size` is always set identically to `(800,800)` in `SlicePlot` and the like.

This can be overcome by doing a `set_buff_size` call, but this call is also somewhat counterintuitive, as it is in reverse order (height, width).  I think, but am not certain, that we can get desired behavior by setting `buff_size` to `None` in the constructor, and then inside `_setup_plots` (where we'll have the aspect ratio info as well as the figure size, although we should probably assume 100 or 300 DPI) we can guess the buffer size.

(Thanks to @maxhutch for the pointer on this.)

----------------------------------------
- Bitbucket: https://bitbucket.org/yt_analysis/yt/issue/1021

**Comment by ngoldbaum** [comment]
This is worth doing still. We should more intelligently choose the image size.

**Comment by sheagk** [comment]
OK, it looks like I should edit `_recreate_frb` in `plot_window.py` to manage the `buff_size` intelligently based on aspect ratio and window size, since I believe both of those are accessible to the class.  Handling the dpi is trickier though, since that won't be specified until `save` (on a related note, can someone point me to the right `save` function?)  I could add an optional argument to `recreate_frb` to allow it to take in either the dpi or (more generally) mpl_kwargs.  Then, when calling `save`, it could check whether the dpi is specified.  If it is, then recreate the buffer with the right size (if necessary); if it isn't, then set the dpi in `save` based on the window size and `buff_size`.

However, all of this is made significantly more complicated by the fact that the buffer sometimes fills the image (i.e. if the colorbar and axes are turned off) and sometimes doesn't.  The former case is easy, but when the buffer doesn't fill the image, then you need to provide extra dots to show all the resolution elements, so you need a higher dpi for a given `buff_size`.  Exactly how much higher may not be trivial -- do you have access to quantities like axes.top and axes.left without first running _setup_plots?

**Comment by ngoldbaum** [comment]
No, we don't create the matplotlib until we actually need to save something to disk or display it. The scale of the axes is known ahead of time (although I think right now we don't actually compute it until the plot is about to be created), see `ImagePlotMPL._get_best_layout`, which calculates the axes extents for the main plot axis and the colorbar axis.
